faf-java-server-app/src/main/java/com/faforever/server/game/SpoofDetectorService.java

```diff
@@ -0,0 +1,118 @@
+import com.faforever.server.entity.AvatarAssociation;
+import com.faforever.server.entity.Game;
+import com.faforever.server.entity.Player;
+import com.faforever.server.entity.Rating;
+import com.faforever.server.error.ErrorCode;
+import com.faforever.server.error.Requests;
+import com.faforever.server.player.PlayerService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * When players joins a game, they send their own information (like name and rating) to other players. This information
+ * can, therefore, not be trusted. Players can send the information they received from others to this service in order
+ * to have it verified. If this service receives multiple reports of incorrect data, it will tell all peers to
+ * disconnect the player who spoofs its data.
+ */
+@Service
+@Slf4j
+public class SpoofDetectorService {
+
+  private final PlayerService playerService;
+  private final ClientService clientService;
+
+  public SpoofDetectorService(PlayerService playerService, ClientService clientService) {
+    this.playerService = playerService;
+    this.clientService = clientService;
+  }
+
+  public void verifyPlayer(Player reporter, int reporteeId, String name, float mean, float deviation, String country, String avatarUrl, String avatarDescription) {
+    Game reporterGame = reporter.getCurrentGame();
+    Requests.verify(reporterGame != null, ErrorCode.THIS_PLAYER_NOT_IN_GAME);
+
+    Optional<Player> reporteeOptional = playerService.getOnlinePlayer(reporteeId);
+    Requests.verify(reporteeOptional.isPresent(), ErrorCode.PLAYER_NOT_ONLINE, reporteeId);
+    Player reportee = reporteeOptional.get();
+
+    Game reporteeGame = reportee.getCurrentGame();
+    Requests.verify(reporteeGame != null, ErrorCode.OTHER_PLAYER_NOT_IN_GAME, reporteeId);
+
+    Requests.verify(reporterGame.equals(reporteeGame), ErrorCode.NOT_SAME_GAME, reportee);
+
+    boolean passesValidation = verifyName(reporter, name, reportee)
+      && verifyRating(reporter, mean, deviation, reportee)
+      && verifyCountry(reporter, country, reportee)
+      && verifyAvatar(reporter, avatarUrl, avatarDescription, reportee);
+
+    if (!passesValidation) {
+      reportee.getFraudReporterIds().add(reporter.getId());
+
+      int reportedFrauds = reportee.getFraudReporterIds().size();
+      if (reportedFrauds > 1 && reportedFrauds >= reporterGame.getConnectedPlayers().size() / 2) {
+        Collection<Player> peers = reporteeGame.getConnectedPlayers().values().stream()
+          .filter(player -> !Objects.equals(player.getId(), reporteeId))
+          .collect(Collectors.toList());
+
+        clientService.disconnectPlayerFromGame(reporteeId, peers);
+      }
+    }
+  }
+
+  private boolean verifyAvatar(Player reporter, String avatarUrl, String avatarDescription, Player reportee) {
```

Comment:
Maybe just nitpicking but I would expect method verifyXy to be void and throw exception if something is not valid. If it returns boolean maybe hasValidAvatar would be better.

Reply:
Not nitpicking for me, but totally valid :) I adjusted it.

I don't throw an exception because exceptions should not be used as flow control in non-exceptional cases. Also, this way we could easily switch to `&` instead of `&&` in case we want to run and log all verifications instead of stopping verification at the first mismatch.